File: 2020/9.rb
lines = open("9.txt").read.split("\n").map(&:to_i)
curr = []
idx = 0
while true
	if idx >= 25
		num = lines[idx]
		works = curr.combination(2).to_a.select {|x| x.sum == num}.size() >= 1
		break if !works
		curr.shift
		curr << num
	else
		curr << lines[idx]
	end
	idx += 1
end
puts num

idx2 = 0
sum = 0
len = 0
while true
	while sum < num
		sum += lines[idx2 + len]
		len += 1
	end
	break if sum == num
	sum -= lines[idx2]
	idx2 += 1
	len -= 1
end
temp = lines[idx2..idx2 + len - 1].sort
puts temp[0] + temp[-1]

File: 2020/9.py
nums = list(map(int, open('9.txt').read().split('\n')))

def sum_to(preamble, num):
    for i in range(len(preamble)):
        for j in range(i + 1, len(preamble)):
            if preamble[i] + preamble[j] == num: return True
    return False

idx = 25
while idx < len(nums):
    if not sum_to(nums[idx-25:idx], nums[idx]): break
    idx += 1

invalid_number = nums[idx]
# part 1: pretty easy, did it pretty quickly
print (invalid_number)

window = 1
idx = 1
window_arr = []
while idx + window < len(nums):
    while sum(nums[idx:idx + window + 1]) < invalid_number:
        window += 1
    if sum(nums[idx:idx + window + 1]) == invalid_number:
        window_arr = nums[idx:idx + window + 1]
        break
    idx += 1
    window = 1

# part 2: also did this one quickly
print (max(window_arr) + min(window_arr))